docs/DevGuide/TensorExpressions.md

```diff
@@ -0,0 +1,411 @@
+index order used for the LHS tensor must be the same in the LHS and RHS, e.g.
+the following is not allowed and will yield a runtime error:
+```
+// ERROR: index order for L on LHS and RHS is not the same
+tenex::update<ti::a, ti::b>(
+    make_not_null(&L), L(ti::b, ti::a) + 2.0 * S(ti::b, ti::a));
+```
+
+**Note:** It is not advised to use very large RHS expressions with
+\ref tenex::update "update" because runtime performance does not scale well as
+the number of operations gets very large. This is because
+\ref tenex::evaluate "evaluate" breaks up large expressions into smaller ones,
+but \ref tenex::update "update" cannot. One way around this is to break up the
+expression and use more than one call to \ref tenex::update "update".
+
+# Compile time math checks {#te_compile_time_math_checks}
+For all operations, mathematical legality is checked at compile time. The
+compiler will catch what is not sound to write on paper, which includes things
+like no repeated indices, can't divide by a tensor with rank > 0, and that
+spatial dimensions, frames, valences, index types (spatial or spacetime), and
+ranks of tensors match where they should.
+
+Here are some examples of illegal math that the compiler will catch:
+
+```
+tnsr::ab<double, 3, Frame::Inertial> R{};
+tnsr::ab<double, 3, Frame::Inertial> S{};
+tnsr::ab<double, 3, Frame::Grid> T{};
+tnsr::AB<double, 2, Frame::Inertial> G{};
+
+// ERROR: LHS and RHS indices don't match
+auto result1 = tenex::evaluate<ti::a, ti::c>(R(ti::a, ti::b) + S(ti:a::ti::b));
+// ERROR: Can't add Tensors with different indices
+auto result2 = tenex::evaluate<ti::a, ti::b>(R(ti::a, ti::b) + S(ti:a::ti:c));
+// ERROR: Repeated index in the RHS
+auto result3 =
+    tenex::evaluate<ti::a, ti::b, ti::c>(R(ti::a, ti::b) * S(ti:a::ti::c));
+// ERROR: Can't add Tensors with different Frame types
+auto result4 = tenex::evaluate<ti::a, ti::b>(R(ti::a, ti::b) + T(ti:a::ti::b));
+// ERROR: Can't contract indices with different number of spatial dimensions
+auto result5 = tenex::evaluate(R(ti::a, ti::b) * G(ti::A, ti::B));
+// ERROR: Can't divide by a rank > 0 Tensor
+auto result6 = tenex::evaluate<ti::a, ti::b>(R(ti::a, ti::b) / S(ti::a, ti::b));
+```
+
+# Support for data types and operations {#te_data_type_and_op_support}
+
+## Data types {#te_data_type_support}
+The RHS expression may contain a mixture of number terms and `Tensor` terms,
+e.g. `0.5 * T(ti::a)`.
+
+Currently supported data types for number terms:
+- `double`
+- `std::complex<double>`
+
+Currently supported underlying data types for `Tensor` terms:
+- `double`
+- `std::complex<double>`
+- `DataVector`
+- `ComplexDataVector`
+
+Support for more types can be added.
+
+## Operations {#te_operation_support}
+It's possible for terms in the expression to have different data types. The
+following table shows the data type that results from performing a binary
+operation (`+`, `-`, `*`, `/`) between two terms of given data types:
+
+<table>
```

Comment:
It could be that the vscode preview is a little off, but I'm a bit confused about the labels on the table. If this is supposed to show the resulting type when doing operations on allowed types, aren't the outside operand types both RHS and the resulting type in the middle of the table the LHS? Here's a png of my preview, so it might just be that that's off. ![TE_Preview](https://github.com/sxs-collaboration/spectre/assets/47013010/e58fccd4-0b91-4bc0-99ad-1607c4e62636)

Reply:
For me it renders differently when I open the documentation locally in a browser, so I would try that! This is what the table looks like for me in a browser (and it has horizontal scroll):

![image](https://github.com/sxs-collaboration/spectre/assets/22260462/be8a9cb5-e376-4944-899f-8bd0d19c2d79)

Reply:
and I'm actually just now realizing the LHS and RHS operand type row and column headers should not include the first row and column that they do and I'll fix that - what I mean is like "RHS operand type" header should only go as far left as the `double` column header, not also the column before it that includes the LHS operand types. So yah, will fix that

Reply:
Ah okay cool, I think I was just confused on what LHS and RHS meant here, as I assumed that something like the resulting type ex) auto L = tenex::evaluate<ti::a, ti:b:> would be the LHS and the RHS operations were things like what you're multiplying to produce the LHS tensor.

Reply:
Yeah I think fixing the column headers would be clearer.